docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+
+* Local or per-thread memory is read-write storage only visible to the
+  threads defining the given variables. The size of a block for a given kernel,
+  the number of concurrent warps are limited by local memory usage.
+  This relates to an important aspect: occupancy. This is the default memory
+  namespace.
+
+* Shared memory is read-write storage visible to all the threads in a given
+  block.
+
+* Distributed shared memory is read-write storage visible to all the threads
+  in a given block cluster.
+
+* Global memory is read-write storage visible to all threads in a given grid.
+  There are specialized versions of global memory with different usage
+  semantics which are typically backed by the same hardware storing global.
+
+  * Constant memory is read-only storage visible to all threads in a given
+    grid. It is a limited segment of global with queryable size.
+
+  * Texture memory is read-only storage visible to all threads in a given grid
+    and accessible through additional APIs.
+
+  * Surface is a writable version of texture memory.
+
+Execution Model
+===============================================================================
+
+HIP programs consist of two distinct scopes:
+
+* The host-side API running on the host processor. There are to APIs available:
+
+  * The HIP runtime API which enables use of the single-source programming
+    model.
+
+  * The HIP driver API which sits at a lower level and most importantly differs
+    by removing some of the facilities provided by the runtime API, most
+    importantly around kernel launching and argument setting. It is geared
+    towards implementing abstractions atop, such as the runtime API itself.
+
+* The device-side kernels running on GPUs. Both the host and the device-side
+  APIs have synchronous and asynchronous functions in them.
+
+Host-side execution
+-------------------------------------------------------------------------------
+
+The part of the host-side API which deals with device management and their
+queries are synchronous. All asynchronous APIs, such as kernel execution, data
+movement and potentially data allocation/freeing all happen in the context of
+device streams.
+
+Streams are FIFO buffers of commands to execute relating to a given device.
+Commands which enqueue tasks on a stream all return promptly and the command is
+executed asynchronously. All side-effects of a command on a stream are visible
+to all subsequent commands on the same stream. Multiple streams may point to
+the same device and those streams may be fed from multiple concurrent host-side
+threads. Execution on multiple streams may be concurrent but isn't required to
+be.
+
+Asynchronous APIs involving a stream all return a stream event which may be
+used to synchronize the execution of multiple streams. A user may enqueue a
+barrier onto a stream referencing an event. The barrier will will block until
+the command related to the event does not complete, at which point all
```

Comment:
Shouldn't it be waiting until it completes?

```suggestion
the command related to the event does complete, at which point all
```

Reply:
Ok. Just realized it should also be "all the command`s`", just so you don't overlook it when fixing it

Reply:
Fixed.